Handle Notification.close in webpushd

beidson opened the pull request:
#### 0794e187a8ecab72143d0a90c889977676da3f86
<pre>
Handle Notification.close in webpushd
<a href="https://rdar.apple.com/131363011">rdar://131363011</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=277171">https://bugs.webkit.org/show_bug.cgi?id=277171</a>

Reviewed by Chris Dumez.

IPC message plumbing and a little new code inside webpushd itself.

* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::cancelNotification):
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::cancelNotification):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::cancelNotification):
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.h:
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.mm:
(-[_WKMockUserNotificationCenter removePendingNotificationRequestsWithIdentifiers:]):
(-[_WKMockUserNotificationCenter removeDeliveredNotificationsWithIdentifiers:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/281438@main">https://commits.webkit.org/281438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1fc614a3e25bf636c2bcac6bb4194693104c1404

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12427 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10425 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62026 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10655 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48549 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7275 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36620 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51862 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29393 "Found 7 new API test failures: /WPE/TestConsoleMessage:/webkit/WebKitConsoleMessage/js-exception, /WPE/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WPE/TestConsoleMessage:/webkit/WebKitConsoleMessage/security-error, /WPE/TestLoaderClient:/webkit/WebKitWebView/load-request, /WPE/TestWebKitFindController:/webkit/WebKitFindController/max-match-count, /WPE/TestWebKitFindController:/webkit/WebKitFindController/match-count, /WPE/TestResources:/webkit/WebKitWebResource/get-data-empty (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33326 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9100 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9346 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65554 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3827 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55898 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51842 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/56051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/13283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3194 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35058 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36140 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37228 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -705,6 +705,15 @@ static bool platformShouldPlaySound(const WebCore::NotificationData& data)
 
+void WebPushDaemon::cancelNotification(PushClientConnection& connection, const WTF::UUID& notificationID)
+{
+    NSString *placeholderBundleIdentifier = platformNotificationCenterBundleIdentifier(connection);
```
May be nice to use a RetainPtr here. We try and be extra safe nowadays.

cdumez commented on Source/WebKit/webpushd/WebPushDaemon.mm:
```diff
@@ -705,6 +705,15 @@ static bool platformShouldPlaySound(const WebCore::NotificationData& data)
+    NSString *placeholderBundleIdentifier = platformNotificationCenterBundleIdentifier(connection);
+    RetainPtr center = adoptNS([[m_userNotificationCenterClass alloc] initWithBundleIdentifier:placeholderBundleIdentifier]);
+
+    [center removePendingNotificationRequestsWithIdentifiers:@[ (NSString *)notificationID.toString() ]];
```
`(NSString *)notificationID.toString()` is not particularly cheap. We may want to cache the result given that it is used twice.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
```diff
@@ -509,21 +520,28 @@ function notificationToString(n)
             port.postMessage("getNotifications failed: " + exception);
         });
+    } else if (message === "closeAllNotifications") {
+
```
We're not doing anything here? Maybe worth a comment?

cdumez reviewed (approved):
LGTM % nits

beidson committed:
Handle Notification.close in webpushd
rdar://131363011
https://bugs.webkit.org/show_bug.cgi?id=277171

Reviewed by Chris Dumez.

IPC message plumbing and a little new code inside webpushd itself.

* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::cancelNotification):
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.messages.in:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::cancelNotification):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::cancelNotification):
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.h:
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.mm:
(-[_WKMockUserNotificationCenter removePendingNotificationRequestsWithIdentifiers:]):
(-[_WKMockUserNotificationCenter removeDeliveredNotificationsWithIdentifiers:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: https://commits.webkit.org/281438@main
